verification/policy.go

```diff
@@ -0,0 +1,153 @@
+}
+
+type TrustPolicy struct {
+	// Name of the policy statement
+	Name string `json:"name"`
+	// RegistryScopes that this policy statement affects
+	RegistryScopes []string `json:"registryScopes"`
+	// SignatureVerification setting for this policy statement
+	SignatureVerification string `json:"signatureVerification"`
+	// TrustStore this policy stament uses
+	TrustStore string `json:"trustStore,omitempty"`
+	// TrustedIdentities this policy statement pins
+	TrustedIdentities []string `json:"trustedIdentities,omitempty"`
+}
+
+func isPresent(val string, values []string) bool {
+	for _, v := range values {
+		if v == val {
+			return true
+		}
+	}
+	return false
+}
+
+// ValidatePolicyDocument validates a policy document according to it's version's rule set.
+// if any rule is violated, returns an error
+func ValidatePolicyDocument(policyDoc *PolicyDocument) error {
+	// Constants
+	wildcardScope := "*"
+	supportedPolicyVersions := []string{"1.0"}
+	supportedVerificationPresets := []string{"strict", "permissive", "audit", "skip"}
+	supportedTrustStorePrefixes := []string{"ca", "signingservice"}
+
+	// Validate Version
+	if !isPresent(policyDoc.Version, supportedPolicyVersions) {
+		return errors.New("Version '" + policyDoc.Version + "' is not supported")
+	}
+
+	// Validate the policy according to 1.0 rules
+	if len(policyDoc.TrustPolicies) == 0 {
+		return errors.New("Trust Policy document can not have zero statements")
+	}
+	var uniqueStatementNames []string
+	registryScopeCount := make(map[string]int)
+	for _, statement := range policyDoc.TrustPolicies {
+
+		// Verify statement name is valid
+		if statement.Name == "" {
+			return errors.New("Policy statement is missing a name")
+		} else {
+			if !isPresent(statement.Name, uniqueStatementNames) {
+				uniqueStatementNames = append(uniqueStatementNames, statement.Name)
+			}
+		}
+
+		// Verify registry scopes are valid
+		if len(statement.RegistryScopes) == 0 {
+			return errors.New("Policy statement has zero registry scopes")
+		} else {
+			if len(statement.RegistryScopes) > 1 && isPresent(wildcardScope, statement.RegistryScopes) {
+				return errors.New("Wildcard scope can not be shared with other registry scopes")
+			}
+			for _, scope := range statement.RegistryScopes {
+				registryScopeCount[scope]++
+			}
+		}
+
+		// Verify signature verification preset is valid
+		if !isPresent(statement.SignatureVerification, supportedVerificationPresets) {
+			return errors.New("SignatureVerification '" + statement.SignatureVerification + "' is not supported")
+		}
+
+		// strict and permissive verification needs trust store to verify authenticity
+		if statement.SignatureVerification == "strict" || statement.SignatureVerification == "permissive" {
+			if statement.TrustStore == "" {
+				return errors.New("Verification statement with strict or permissive preset is missing a trust store")
+			}
+		}
+
+		// Verify trust store type is valid
+		if statement.TrustStore != "" {
```

Comment:
nit: this is statement is not necessary, if `TrustStore` is empty `strings.Index` will return `-1`.

Reply:
Need to keep this one because we need to differentiate between empty trust store and a trust store with an invalid type. If we remove the `if` statement, then statements without a trust store (valid case for `skip` verification), we would throw *unsupported trust store type* error

Reply:
@rgnote, can you add that feedback to the comment, so the next person understands the reasoning?

Reply:
Will do

Reply:
It would be better to combine this and previous validation as
```
if signatureverification is skip
  truststore and trustedentities must be empty
else 
  both should have valid values
```

Reply:
The code becomes more complex that way
```
if signatureverification is skip
  if truststore and trustedentities are not empty
    throw error
else
  if truststore is empty
    throw error
  if truststore is not valid
    throw error
  if trustedidentities are zero
    throw error
  if trustedidentities are invalid
    throw error
```

Reply:
Agreed, but it can me made to read easier by moving trustStore, trustIdentity validation as separate methods. The current logic is more difficult to follow, we've missed the case that skip must not specify trustStore and trustedIdentity, it could be kept optional but that is confusing to users. Can be addressed in next PR.